### PR TITLE
Added MySQL Support for DB options

### PR DESCRIPTION
**Pull request by DevelopIdeas**
For this like myself who already have systems running with a MySQL db we'd like the Frigate data in

### CODE REVIEW

File: frigate/config.py [hunk 1]
```diff
@@ -644,12 +644,36 @@ def _get_ffmpeg_cmd(self, ffmpeg_input: CameraInput):
...
+        default="localhost", title="Database host."
+    )
+    user: str = Field(
+        default="frigate", title="Database user."
+    )
+    password: str = Field(
+        default="frigate", title="Database password."
+    )
+    port: int = Field(
+        default=3306, title="Database password."
```
Comment (NickM-27) [inline frigate/config.py:675]:
```suggestion
        default=3306, title="Database server port."
```

File: frigate/models.py [hunk 2]
```diff
@@ -1,14 +1,15 @@
...
+from playhouse.mysql_ext import *
 
 
 class Event(Model):
     id = CharField(null=False, primary_key=True, max_length=30)
     label = CharField(index=True, max_length=20)
     camera = CharField(index=True, max_length=20)
     start_time = DateTimeField()
-    end_time = DateTimeField()
+    end_time = DateTimeField(null=True)
```
Comment (NickM-27) [inline frigate/models.py:12]: Just curious, why is this needed?

**Review by NickM-27 — CHANGES_REQUESTED** [review]

**Comment by NickM-27** [comment]
Thanks for your contribution!

File: frigate/models.py (hunk 2 above)
Comment (DevelopIdeas) [reply frigate/models.py:12]: I found when testing, pymysql doesnt set DEFAULT NULL without specifying this.
frigate/events.py tries to set this as None when starting an event.

File: docker/Dockerfile.wheels [hunk 3]
```diff
@@ -34,7 +34,8 @@ RUN pip3 wheel --wheel-dir=/wheels \
     matplotlib \
     click \
     setproctitle \
-    peewee
+    peewee \
+    pymysql
```
[the remark this answers is not in the document]
Comment (DevelopIdeas) [reply docker/Dockerfile.wheels:38]: pymysql is required by peewee when connecting to a MySQL DB, without this installed an Exception is raised

[the remark this answers is not in the document]
Comment (NickM-27) [reply docker/Dockerfile.wheels:38]: interesting, good to know, I figured they would have bundled it but that makes sense 👍

File: frigate/app.py [hunk 4]
```diff
@@ -33,6 +35,8 @@
 
 logger = logging.getLogger(__name__)
 
+class ReconnectMySQLDatabase(ReconnectMixin, MySQLDatabase):
```
[the remark this answers is not in the document]
Comment (DevelopIdeas) [reply frigate/app.py:38]: Sure, I can move the logic from frigate/app.py init_database to a database.py along with the import and recommit if you like?

[the remark this answers is not in the document]
Comment (DevelopIdeas) [reply frigate/app.py]: I've done the refactor to database.py, do i need to create a new pull request to merge the changes?

[the remark this answers is not in the document]
Comment (NickM-27) [reply frigate/app.py]: They're all contained here so it's good as far as that goes 👍

[the remark this answers is not in the document]
Comment (DevelopIdeas) [reply frigate/app.py]: Ah great! 👍

**Comment by blakeblackshear** [comment]
I'm not sure I want to add the overhead of compatibility with multiple database options.

**Comment by DevelopIdeas** [comment]
> I'm not sure I want to add the overhead of compatibility with multiple database options.

I've got another commit relating to this.
Should I send it over or not bother if you're considering not adding support?

**Comment by blakeblackshear** [comment]
If was 100% plug and play, then I would consider it. There are a few places where I use custom sql queries, and I am not sure they will work with anything other than sqlite. I just frankly don't have the bandwidth to commit to testing multiple databases and maintaining any db specific workarounds for each release. If you were willing to commit to maintaining it and including a robust test suite alongside this, then maybe. I would likely just remove support as soon as it became a burden for me.

**Comment by onedr0p** [comment]
Personally I would rather see support using an ORM (like SQLAlchemy since this is python) rather than implementing a single database type. Postgres is a much more mature database over mariadb/mysql but it would give people options to support either one in addition to sqlite.

**Comment by luisgreen** [comment]
> I'm not sure I want to add the overhead of compatibility with multiple database options.

Hi @blakeblackshear I've detected a race condition, I have 2 cameras that overlaps a zone, then they both detects the same event, sometimes only ONE gets recorded even if it where detected. I know they where detected because the MQTT event arrives my HA installation. Also, when a single camera gets two events, where two objeccts are too close, coincidentially when a person is strolling a dog... Sometimes it get the dog, sometimes it get the person.... But not both as it should. BUT both are detected for a tiny space of time. I'm not reading the full code, but it could be possible the database locking is preventing multiple transactions? If thats the case. Maybe you should consider moving towards a more robust DB.

Any volunteer to share the maintenance of a RDBMS with me?

**Comment by NickM-27** [comment]
>  I have 2 cameras that overlaps a zone, then they both detects the same event, sometimes only ONE gets recorded even if it where detected. I know they where detected because the MQTT event arrives my HA installation. Also, when a single camera gets two events, where two objeccts are too close, coincidentially when a person is strolling a dog... Sometimes it get the dog, sometimes it get the person.... But not both as it should.

to be honest I doubt this has anything to do with the database. I've had scenarios like this plenty of times and I've never seen any issue with it recording all events to the DB. Without seeing your config, the MQTT payloads, etc. there's no way to know this is related to the DB at all. 

> I'm not reading the full code, but it could be possible the database locking is preventing multiple transactions?

I'd suggest doing that because it seems the way you're assuming it works is incorrect.

**Comment by blakeblackshear** [comment]
@luisgreen what you are describing doesn't sound like a race condition or database lock issue to me. I don't think this closed PR has anything to do with what you are seeing.

If you want help troubleshooting what you are seeing, I suggest opening an issue and providing the requested information.

**Comment by jangrewe** [comment]
Could this be re-opened? I'd really love to see an ORM implementation, where i could just plug in any SQL URL and let the library handle the connection and queries, because let's be honest: SQLite sucks. ;-)
